src/events.jl

```diff
@@ -1,85 +1,210 @@
+    FITSMetadata{H}
 
-A structure containing metadata from FITS file headers.
+Metadata associated with a FITS or events file (`.fits` or `.evt`).
 
-## Fields
+$(FIELDS)
+"""
+struct FITSMetadata{H}
+    "Path to the FITS file."
+    filepath::String
+    "HDU index that the metadata was read from."
+    hdu::Int
+    "Units of energy (currently just ENERGY or PI or HPA)"
+    energy_units::Union{Nothing,String}
+    "Extra columns that were requested during read."
+    extra_columns::Dict{String,Vector}
+    "Fits headers from the selected HDU."
+    headers::H
+end
+
+function Base.show(io::IO, ::MIME"text/plain", @nospecialize(m::FITSMetadata))
+    println(
+        io,
+        "FITSMetadata for $(m.filepath)[$(m.hdu)] with $(length(m.extra_columns)) extra column(s)",
+    )
+end
+
+"""
+    EventList{TimeType, MetaType <: FITSMetadata}
+
+Container for an events list. Generally should not be directly constructed, but
+read from file using [`readevents`](@ref).
 
-- `headers::Vector{Dict{String,Any}}`: A vector of dictionaries containing header information from each HDU.
+$(FIELDS)
 """
-struct DictMetadata
-    headers::Vector{Dict{String,Any}}
+struct EventList{TimeType<:AbstractVector,MetaType<:FITSMetadata}
+    "Vector with recorded times."
+    times::TimeType
+    "Vector with recorded energies (else `nothing`)"
+    energies::Union{Nothing,TimeType}
+    meta::MetaType
 end
 
+function Base.show(io::IO, ::MIME"text/plain", @nospecialize(ev::EventList))
+    print(io, "EventList with $(length(ev.times)) times")
+    if !isnothing(ev.energies)
+        print(io, " and energies")
+    end
+    println(io)
+end
+
+"""
+    filter_time!(f, ev::EventList)
+
+Filter all columns of the eventlist based on a predicate `f` applied to the
+times.
+
+# Example
+
+```julia
+# filter only positive times
+filter_time!(t -> t > 0, ev)
+```
+
+See also [`filter_energy!`](@ref).
 """
-    EventList{T}
+filter_time!(f, ev::EventList) = filter_on!(f, ev.times, ev)
 
-A structure containing event data from a FITS file.
+"""
+    filter_energy!(f, ev::EventList)
 
-## Fields
+Filter all columns of the eventlist based on a predicate `f` applied to the
+energies.
 
-- `filename::String`: Path to the source FITS file.
-- `times::Vector{T}`: Vector of event times.
-- `energies::Vector{T}`: Vector of event energies.
-- `metadata::DictMetadata`: Metadata information extracted from the FITS file headers.
+See also [`filter_time!`](@ref).
 """
-struct EventList{T}
-    filename::String
-    times::Vector{T}
-    energies::Vector{T}
-    metadata::DictMetadata
+function filter_energy!(f, ev::EventList)
+    @assert !isnothing(ev.energies) "No energies present in the EventList."
+    filter_on!(f, ev.energies, ev)
+end
+
+function filter_on!(f, src_col::AbstractVector, ev::EventList)
+    @assert size(src_col) == size(ev.times)
+
+    # modified from the Base.filter! implementation
+    j = firstindex(ev.times)
+
+    for i in eachindex(ev.times)
```

Comment:
What does this do?

Reply:
`eachindex` gives you a lazy iterator that goes over all of the indices of an array. If your array has unusual indexing (e.g. starts at 0 or something else, or maybe uses `IndexCartesian`), it gives you something that will hit every element.

In this case, it's equivalent to `firstindex(ev.times):lastindex(ev.times)`, or equivalently `1:length(ev.times)`, but would work with other array types just fine.

[Somewhere in the Julia docs](https://docs.julialang.org/en/v1/manual/arrays/#Linear-indexing) it says to prefer `eachindex` over `1:length` so that code is more portable. In this case it really doesn't make a difference. So many ways to skin a cat!